xdsl/irdl/constraints.py

```diff
@@ -351,31 +351,34 @@ class ParamAttrConstraint(
     and also constrain its parameters with additional constraints.
     """
 
-    base_attr: type[ParametrizedAttributeCovT]
-    """The base attribute type."""
+    base_constr: BaseAttr[ParametrizedAttributeCovT]
+    """The base attribute type constraint."""
 
     param_constrs: tuple[GenericAttrConstraint[Attribute], ...]
     """The attribute parameter constraints"""
 
     def __init__(
         self,
-        base_attr: type[ParametrizedAttributeCovT],
+        base_attr: type[ParametrizedAttributeCovT]
+        | BaseAttr[ParametrizedAttributeCovT],
         param_constrs: Sequence[
             (Attribute | type[Attribute] | GenericAttrConstraint[Attribute])
         ],
     ):
+        if not isinstance(base_attr, BaseAttr):
+            base_attr = BaseAttr(base_attr)
         constrs = tuple(attr_constr_coercion(constr) for constr in param_constrs)
-        object.__setattr__(self, "base_attr", base_attr)
+        object.__setattr__(self, "base_constr", base_attr)
         object.__setattr__(self, "param_constrs", constrs)
 
     def verify(
         self,
         attr: Attribute,
         constraint_context: ConstraintContext,
     ) -> None:
-        if not isinstance(attr, self.base_attr):
+        if not isinstance(attr, self.base_constr.attr):
```

Comment:
still isinstance instead of `.verify` because that function by itself does not communicate to the type system that it raises an error if the types aren't right.